From 3ce3e14cd42ca04d2dd714d17a82285a5e85a917 Mon Sep 17 00:00:00 2001
From: Jiri Olsa <jolsa@redhat.com>
Date: Tue, 19 Aug 2014 15:27:03 -0400
Subject: [tools] perf/top: Use parse_options_usage() for -s option failure

Message-id: <1408462094-14194-255-git-send-email-jolsa@redhat.com>
Patchwork-id: 88243
O-Subject: [PATCH RHEL7.1 BZ1131394 254/325] perf top: Use parse_options_usage() for -s option failure
Bugzilla: 1131394
RH-Acked-by: Prarit Bhargava <prarit@redhat.com>
RH-Acked-by: Tony Camuso <tcamuso@redhat.com>

From: Jiri Olsa <jolsa@kernel.org>

Bugzilla: 1131394
https://bugzilla.redhat.com/show_bug.cgi?id=1131394

upstream
========
commit d37a92dcb45094dc02836c8a77c693c6f9916fb2
Author: Namhyung Kim <namhyung.kim@lge.com>
Date: Fri Nov 1 16:33:14 2013 +0900

description
===========
The -s (--sort) option was processed after normal option parsing so that
it cannot call the parse_options_usage() automatically.  Currently it
calls usage_with_options() which shows entire help messages for event
option.  Fix it by showing just -s options.

  $ perf top -s help
    Error: Unknown --sort key: `help'

   usage: perf top [<options>]

      -s, --sort <key[,key2...]>
                            sort by key(s): pid, comm, dso, symbol, ...
---

Signed-off-by: Jarod Wilson <jarod@redhat.com>

diff --git a/tools/perf/builtin-top.c b/tools/perf/builtin-top.c
index 21db76d..ca5ca37 100644
--- a/tools/perf/builtin-top.c
+++ b/tools/perf/builtin-top.c
@@ -1040,7 +1040,7 @@ parse_percent_limit(const struct option *opt, const char *arg,
 
 int cmd_top(int argc, const char **argv, const char *prefix __maybe_unused)
 {
- int status;
+ int status = -1;
  char errbuf[BUFSIZ];
  struct perf_top top = {
   .count_filter      = 5,
@@ -1159,8 +1159,10 @@ int cmd_top(int argc, const char **argv, const char *prefix __maybe_unused)
  if (sort_order == default_sort_order)
   sort_order = "dso,symbol";
 
- if (setup_sorting() < 0)
-  usage_with_options(top_usage, options);
+ if (setup_sorting() < 0) {
+  parse_options_usage(top_usage, options, "s", 1);
+  goto out_delete_evlist;
+ }
 
  /* display thread wants entries to be collapsed in a different tree */
  sort__need_collapse = 1;
-- 
1.7.1